zcash_client_sqlite/src/wallet/db.rs

```diff
@@ -76,6 +76,43 @@ CREATE INDEX "addresses_accounts" ON "addresses" (
     "account_id" ASC
 )"#;
 
+/// Stores ephemeral transparent addresses used for ZIP 320. For each account, these addresses are
+/// allocated sequentially by address index under custom scope 2 at the "change" level of the BIP 32
+/// address hierarchy. Only "reserved" ephemeral addresses, that is addresses that have been allocated
+/// for use in a ZIP 320 transaction proposal, are stored in the table. Addresses are never removed.
+/// New ones should only be reserved via the `WalletWrite::reserve_next_n_ephemeral_addresses` API.
+/// All of the addresses in the table should be scanned for incoming funds.
+///
+/// - The `address` column contains the string (Base58Check) encoding of a transparent P2PKH address.
+/// - The `used_in_tx` column indicates that the address has been used by this wallet in a transaction
+///   (which has not necessarily been mined yet). This should only be set once, when the txid is known.
+/// - If the `mined_in_tx` column is non-null, the address has been observed in a mined transaction
+///   (which may have been sent by this wallet or another one using the same seed, or by a TEX address
+///   recipient sending back the funds). This is used to advance the "gap limit", as well as to
+///   heuristically reduce the chance of address reuse collisions with another wallet using the same
+///   seed.
+///
+/// Note that the fact that `used_in_tx` and `mined_in_tx` reference specific transactions is primarily
+/// a debugging aid. We only really care which addresses have been used, and whether we can allocate
+/// a new address within the gap limit.
```

Comment:
What happens when a reorg occurs? Specifically, what should the implementation of `WalletWrite::truncate_to_height` do to this table? I think the answer is "once a transaction is detected and linked into either column, do not remove that link even if the transaction becomes unmined and expires due to a reorg". So even though this is a debugging aid, I think it is required that these be tracked correctly.

This would also address the other issue I saw, which is "what if it is used in multiple transactions". If the intent is that these columns are write-only for a given wallet, then it doesn't matter if we see it multiple times; once is enough.

Maybe the thing that is a debugging aid here is the _distinction_ between `used_in_tx` and `mined_in_tx`. If we didn't need that debugging aid, we would still I think need to permanently record the first time we observe any usage.

Reply:
If a reorg occurs, `mined_in_tx` should be set to NULL and `used_in_tx` should not. `used_in_tx` is set before the transaction is mined in any case, so there is no reason to reset it. I will make that change.

Reply:
It's actually better for the query in `reserve_next_n_ephemeral_addresses` that uses `mined_in_tx` to join on the transactions table and restrict to indices for which `t.mined_height IS NOT NULL`. This avoids forgetting that the address had been observed as mined in `wallet::transparent::ephemeral_address_check_internal`, which uses `COALESCE(used_in_tx, mined_in_tx)`. I will document why we're doing that.